// poemd-parser/lib/message.js
"use strict";
exports.__esModule = true;
exports.ONE_OF_EXPECTED = exports.NUMERIC_VALUE_EXPECTED = exports.NO_HEADER_FOUND = exports.UNUSED_KEY = exports.Message = void 0;
var utils_1 = require("./utils");
var MessageLevel;
(function (MessageLevel) {
    MessageLevel["Info"] = "info";
    MessageLevel["Warning"] = "warning";
    MessageLevel["Error"] = "error";
})(MessageLevel || (MessageLevel = {}));
var MessageTemplate = /** @class */ (function () {
    function MessageTemplate(level, format) {
        this.id = MessageTemplate.ID++;
        this.level = level;
        this.format = format;
    }
    MessageTemplate.prototype.specify = function (args) {
        if (args === void 0) { args = {}; }
        return new Message(this, args);
    };
    MessageTemplate.ID = 0;
    return MessageTemplate;
}());
var Message = /** @class */ (function () {
    function Message(template, args) {
        this.template = template;
        this.args = args;
    }
    Message.prototype.toString = function () {
        return this.template.level + " (" + this.template.id + "): " + utils_1.formatString(this.template.format, this.args);
    };
    return Message;
}());
exports.Message = Message;
exports.UNUSED_KEY = new MessageTemplate(MessageLevel.Warning, 'unused key: "{key}"');
exports.NO_HEADER_FOUND = new MessageTemplate(MessageLevel.Info, 'no header found');
exports.NUMERIC_VALUE_EXPECTED = new MessageTemplate(MessageLevel.Error, 'property "{key}" has to be a number (got "{val}"); falling back to default');
exports.ONE_OF_EXPECTED = new MessageTemplate(MessageLevel.Error, 'property "{key}" has to be one of [{vals}] (got "{val}"); falling back to default');
